### PR TITLE
Upstream merge 4 [fixed]

### PR DESCRIPTION
**Pull request by camthegeek**
Following @stoffu's last PR format.

This pull request merges upstream patches after [monero-project#4129](https://github.com/monero-project/monero/pull/4129) which is the last one included in #90. This PR includes upstream PRs from [monero-project#3528](https://github.com/monero-project/monero/pull/3528) merged on August 15 up to [monero-project#4223](https://github.com/monero-project/monero/pull/4223) merged on August 23:

* monero-project#3528
* monero-project#4020
* monero-project#4024
* monero-project#4089
* monero-project#4090
* monero-project#4090
* monero-project#4103
* monero-project#4106
* monero-project#4136
* monero-project#4137
* monero-project#4137
* monero-project#4138
* monero-project#4138
* monero-project#4148
* monero-project#4150
* monero-project#4153
* monero-project#4158
* monero-project#4161
* monero-project#4164
* monero-project#4165
* monero-project#4166
* monero-project#4170
* monero-project#4171
* monero-project#4172
* monero-project#4173
* monero-project#4174
* monero-project#4177
* monero-project#4186
* monero-project#4188
* monero-project#4192
* monero-project#4193
* monero-project#4194
* monero-project#4195
* monero-project#4196
* monero-project#4191
* monero-project#4131
* monero-project#4131
* monero-project#4131
* monero-project#4131
* monero-project#4179
* monero-project#4200
* monero-project#4202
* monero-project#4204
* monero-project#4207
* monero-project#4210
* monero-project#4224
* monero-project#4225
* monero-project#4231
* monero-project#4240
* monero-project#4242
* monero-project#4247
* monero-project#4249
* monero-project#3999
* monero-project#4223

Notable updates:
**Wallet RPC** new commands
* [refresh](https://www.getmonero.org/resources/developer-guides/wallet-rpc.html#refresh)
* [get_address_index](https://www.getmonero.org/resources/developer-guides/wallet-rpc.html#get_address_index)

**Wallet CLI**
* 4188 allows users to import tx secret keys after restoring a wallet that was originally created using 3rd party wallet services such as web wallets so that tx proofs can then be generated.

This group of upstream merges contains many fixes that get us ready for the next patch that will contain many more major updates to AEON.

### CODE REVIEW

**Comment by 420coupe** [comment]
Built & tested on OS Ubuntu 18.10. 

Tested Daemon/Wallet commands.
No issues discovered.

**Comment by NewEraCracker** [comment]
I merged this (#100) and #102 in a forked test branch and built from source.
GUI is not able to generate new wallets (Windows 7)
GUI > English US > Create new wallet > Seed text area is empty.

Tree with merged PRs: Doesn't work:
https://github.com/NewEraCracker/aeon-gui/tree/9c0ddc5df8afe63ebf584fc21d400dcad0022dc4
https://github.com/NewEraCracker/aeon/tree/49c1a4b7886bfe889bdb18a5cd3c99a20222056d

Tree without merged PRs: Works fine:
https://github.com/NewEraCracker/aeon-gui/tree/f1a37baa2d0bb6ca4032b5dc8d9f5ee11f3a5b14
https://github.com/NewEraCracker/aeon/tree/01e5cde59e66f9c9c104e735b32e2433ae7ca9d1

**Comment by BigslimVdub** [comment]
Built and tested windows 10 daemon and wallet CLI no issues basic testing.

**Review by stoffu — COMMENTED** [review]
Please add /monero#4020 to the commit message (helpful for easily locating merged PRs)

**Comment by camthegeek** [comment]
> Please add /monero#4020 to the commit message (helpful for easily locating merged PRs)

It was on 2nd line somehow. Moved to first line.

**Comment by BigslimVdub** [comment]
Build successful on OSX 10.13.6 High sierra with PR 100 and cherry pick #102. Daemon sync from 0 and wallet sync ok. No issues on basic commands Daemon or CLI wallet. Restore from seed ok. Daemon works with standard DB location along with --data-dir usage.

**Comment by BigslimVdub** [comment]
I think once we get this merged we can commit 12.9 as a release candidate @smooth

**Comment by stoffu** [comment]
I don’t think this PR should go into the next point release assuming the release coming soon. It needs a bit more testing period.

**Comment by BigslimVdub** [comment]
```This group of upstream merges contains many fixes that get us ready for the next patch that will contain many more major updates to AEON.```

These changes were pretty well tested with pr99 before the additional few changes were added to create pr100. From the above comment I would believe we would want these changes in 12.9 to set the floor for 13.0. Some members have been running this Pr for a week or more in pools or other nodes with no issues reported. 

Thoughts @iamsmooth ?

**Comment by iamsmooth** [comment]
We'll include it in the next release after this one so people can continue testing it.

File: tests/unit_tests/crypto.cpp [hunk 1]
```diff
@@ -74,17 +74,10 @@ TEST(Crypto, Ostream)
...
-  unsigned char k0[32] = {0}, k1[32] = {0};
-  for (unsigned int i0 = 0; i0 < 256; ++i0)
-  {
-    k0[0] = i0;
-    for (unsigned int i1 = 0; i1 < 256; ++i1)
-    {
-      k1[0] = i1;
-      ASSERT_EQ(!crypto_verify_32(k0, k1), i0 == i1);
-    }
-  }
```
Comment (stoffu) [inline tests/unit_tests/crypto.cpp]: Please look at the code changes carefully. This patch is only meant to add a new test `null_keys`, not to remove any existing ones.

**Comment by stoffu** [comment]
@camthegeek 

The ordering of commits are a bit shuffled around 4136, 4137, 4138, etc. I assume this is Git's failure to handle many consecutive commits being added quickly. When you add many commits to a branch (such as when rebasing), be sure to pause a bit between `git commit` command to ensure that the underlying database gets updated properly.

File: src/crypto/chacha.h [hunk 2]
```diff
@@ -69,22 +69,26 @@ namespace crypto {
     chacha20(data, length, key.data(), reinterpret_cast<const uint8_t*>(&iv), cipher);
   }
 
-  inline void generate_chacha_key(const void *data, size_t size, chacha_key& key) {
+  inline void generate_chacha_key(const void *data, size_t size, chacha_key& key, uint64_t kdf_rounds) {
     static_assert(sizeof(chacha_key) <= sizeof(hash), "Size of hash must be at least that of chacha_key");
     tools::scrubbed_arr<char, HASH_SIZE> pwd_hash;
     crypto::cn_slow_hash(data, size, pwd_hash.data(), 0/*light*/, 0/*variant*/, 0/*prehashed*/);
+    for (uint64_t n = 1; n < kdf_rounds; ++n)
+      crypto::cn_slow_hash(pwd_hash.data(), pwd_hash.size(), 0/*light*/, pwd_hash.data(), 0/*variant*/, 0/*prehashed*/);
```
Comment (stoffu) [inline src/crypto/chacha.h]: This is a wrong patch which is later fixed in 74757a23980fa484792b4997e48635f2eadd229e. It should not have been introduced here in the first place.

**Comment by camthegeek** [comment]
@stoffu 

It does appear that rebase screwed the order in which the commits were arranged. I am left to assume doing an interactive rebase can change the sort again.

File: src/crypto/chacha.h (hunk 2 above)
Comment (camthegeek) [reply src/crypto/chacha.h]: You are referencing the wrong commit in the highlighted lines. Please see c2fa3e1fc instead. This is also 4103 which is before your referenced 4131.

Comment (stoffu) [reply src/crypto/chacha.h]: I'm not sure what you mean, but the very suggested commit https://github.com/aeonix/aeon/commit/c2fa3e1fc072c069fff5f251ef5a6edcffb6754c is the one that contains this wrong patch.

**Pull request closed without merging**